fendermint/ipc/src/parent.rs

```diff
@@ -0,0 +1,202 @@
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        match self
+            .started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => return Err(anyhow!("already started")),
+        }
+
+        let parent_syncer = self.ipc_agent_proxy.clone();
+        let config = self.config.clone();
+        let cache = self.cache.clone();
+
+        let handle =
+            tokio::spawn(async move { sync_with_parent(config, parent_syncer, cache).await });
+        self.handle = Some(handle);
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent(
+    config: Config,
+    agent_proxy: Arc<AgentProxy>,
+    lock: LockedCache,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval));
+
+    loop {
+        interval.tick().await;
+
+        let hashes = fetch_block_hashes(&config, &agent_proxy, &lock).await?;
+        update_top_down_msgs(&config, &agent_proxy, &lock).await?;
+        update_membership(&config, &agent_proxy, &lock).await?;
+
+        let mut cache = lock.write().unwrap();
+        for r in hashes {
+            cache.block_hash.insert_after_lower_bound(r.0, r.1);
+        }
+    }
+}
+
+async fn update_top_down_msgs(
+    _config: &Config,
+    _agent_proxy: &Arc<AgentProxy>,
+    _lock: &LockedCache,
+) -> anyhow::Result<Vec<CrossMsg>> {
+    Ok(vec![])
+}
+
+async fn update_membership(
+    _config: &Config,
+    _agent_proxy: &Arc<AgentProxy>,
+    _lock: &LockedCache,
+) -> anyhow::Result<ValidatorSet> {
+    Ok(ValidatorSet::default())
+}
+
+async fn fetch_block_hashes(
+    config: &Config,
+    agent_proxy: &Arc<AgentProxy>,
+    lock: &LockedCache,
+) -> anyhow::Result<Vec<(BlockHeight, Vec<u8>)>> {
+    let latest_height = match agent_proxy.get_chain_head_height().await {
+        Ok(h) => h,
+        Err(e) => {
+            tracing::warn!("cannot fetch parent chain head due to {e}");
+
+            // not throw errors, caller will retry
+            return Ok(vec![]);
+        }
+    };
+
+    if latest_height < config.chain_head_delay {
+        tracing::debug!("latest height not more than the chain head delay");
+        return Ok(vec![]);
+    }
+
+    let starting_height = {
+        let cache = lock.read().unwrap();
+        // if cache.latest_height() is None, it means we have not started fetching any heights
+        // we just use the latest height minus a lower bound as the parent view
+        let starting_height = cache.block_hash.upper_bound().unwrap_or(max(
+            1,
+            latest_height.saturating_sub(config.chain_head_lower_bound),
+        ));
+        tracing::debug!("polling parent from {starting_height} till {latest_height}");
+
+        starting_height
+
+        // ensure cache is dropped
+    };
+
+    // FIXME: make the fetching batching and concurrent
+    let mut results = vec![];
+    for h in starting_height..=latest_height {
```

Comment:
So basically this is re-reading all the last 900 block heights? There must be a way to handle re-orgs in a more relaxed way, push notifications, maybe compare the last height read with what you have on record to see if there was a change?